src/Microsoft.AspNetCore.Routing/Template/TemplateBinder.cs

```diff
@@ -375,7 +375,14 @@ public static bool RoutePartsEqual(object a, object b)
             var sa = a as string;
             var sb = b as string;
 
-            if (sa != null && sb != null)
+            // In case of strings, consider empty and null the same.
+            // Since null cannot tell us the type, consider it to be a string if the other value is a string.
+            if ((sa != null && string.IsNullOrEmpty(sa) && sb == null) ||
+                (sb != null && string.IsNullOrEmpty(sb) && sa == null))
```

Comment:
I think this would be a little clearer if you wrote this as `sa == string.Empty...` and so on. That's a simpler form of `sa != null && string.IsNullOrEmpty(sa)` and it's closer to what you actually want.

Reply:
`sa != null && string.IsNullOrEmpty(sa)` -> `string.IsNullOrEmpty(sa)`

cause IsNullOrEmpty is already checking for null.

Reply:
Right. and we specifically need to know if it's **exactly** the empty string.

Reply:
Might be better to write it like so:

```C#
        public static bool RoutePartsEqual(object a, object b)
        {
            if (a is string sa && b is string sb)
            {
                return (string.IsNullOrEmpty(sa) && string.IsNullOrEmpty(sb)) ||
                    string.Equals(sa, sb, StringComparison.OrdinalIgnoreCase);
            }
            else
            {
                if (a != null && b != null)
                {
                    // Explicitly call .Equals() in case it is overridden in the type
                    return a.Equals(b);
                }
                else
                {
                    // At least one of them is null. Return true if they both are
                    return a == b;
                }
            }
        }
```

Reply:
This would have the same bug. `is` will return false when one of them is null.